Affichage apogée et fix du TextItem de l'altitude

BaboucheOne opened the pull request:
+ Affichage de l’apogée.
+ Résolution du bug de l'affichage en temps réel de l'altitude.

BaboucheOne committed:
Update BaseStation/src/ui/altitude_graph.py

Override de la methode 'draw' afin de pouvoir ecrire  sur un TextItem.

Ajout de l'affichage de l'apogee.
Fix du textItem de l'altitude.

BaboucheOne committed:
Update BaseStation/src/ui/altitude_graph.py

 + Correction de l'altitude, maintenant affiche en "ft" au lieu de "m"

+ Mofidication de l'emplacement de deux conditions

BaboucheOne committed:
Update altitude_graph.py

+ Reorganisation
+ Ajout de la fonction pour calculer l apogee

BaboucheOne committed:
Update altitude_graph.py

Une autre methode

BaboucheOne committed:
Update altitude_graph.py

Fix

BaboucheOne committed:
Update altitude_graph.py

+ Mise a jour de la methode 'setApogee'

MaitreMenard commented on BaseStation/src/ui/altitude_graph.py:
```diff
@@ -19,30 +19,71 @@ def __init__(self, parent: QtWidgets.QWidget):
+
+        super(AltitudeGraph, self).paintEvent(ev)
+
+    def set_apogee(self, values: list, current_point: int):
```
La logique permettant de calculer l'apogée ne devrait pas se retrouver dans la couche de UI. Il s'agit d'une logique du domaine d'affaire qui devrait plutôt se retrouver dans la couche du domaine (en dessous du controller). Idéalement, on aurait une classe dont la seule responsabilité est de calculer l'apogée à partir d'une liste de points. Cette classe pourrait être utilisée par le consommateur. La classe _AltitudeGraph_ ne devrait servir qu'à l'affichage d'une liste de points d'altitudes et du point de l'apogée préalablement calculé.

BaboucheOne replied:
Si je comprend bien, je dois créer une classe qui servirais juste a calculer l’apogée. Ensuite cette classe pourras interagir avec le graphique juste pour lui passer des informations tel que l’apogée.

MaitreMenard replied:
C'est presque ça, la classe servant à calculer l'apogée va être utilisée par le _Consumer_ et le _Controller_ va pousser les données au graphique, qui va se contenter de l'afficher.

MaitreMenard commented on BaseStation/src/ui/altitude_graph.py:
```diff
@@ -19,30 +19,71 @@ def __init__(self, parent: QtWidgets.QWidget):
+                self.apogee = -1
+                apogee_index = 0
+
+                for i in range(len(values) - 1):
```
Voir commentaire précédent.

MaitreMenard commented:
N'oublie pas de faire des tests unitaires pour valider le bon fonctionnement de ton algorithme de détection d'apogée.

BaboucheOne committed:
Update altitude_graph.py

Ajout de la classe ApogeeCalculator

BaboucheOne committed:
Class update

+ Ajout de tests unitaire
+ Modification de la class ApogeeCalculator

BaboucheOne committed:
Update altitude_graph.py

BaboucheOne committed:
Ajout del'apogee sur le graph

BaboucheOne committed:
Merge branch 'dev' into Affichage-Apogee-et-fix-TextItem-altitude

BaboucheOne committed:
Fix d'Apogee_calculator

+ Reecrire de le class 'ApogeeCalculator'
+ Ajout de test unitaires plus complets
+ Ajout de la methode 'draw_apogee" dans le graphique

BaboucheOne committed:
Mise en forme

+ précision de l'apogee
+ affichage de l'apogee
+ mode replay / temps reel fonctionel
+ mise en forme des fin de fichier

- Bug du a un recule brusque de la bar replay.
- Bug du mode temps reel. [Ne pas afficher dans l'apogee est a 0]

BaboucheOne committed:
Replay bug fix

+ Correction de l'affichage du point d'apogee.

- Bug apparu sur le mode 'real time'

MaitreMenard commented on BaseStation/src/apogee_calculator.py:
```diff
@@ -0,0 +1,43 @@
+        if length == self.last_altitude_index:
+            return
+
+        if self.apogee_index < length or self.last_altitude_index == 0 or (self.last_altitude_index - values[-1]) <= 0 or self.apogee == 0:
```
En générale, j'aime bien extraire les conditions complexes comme celle-là dans une méthode privée dont le nom indique clairement ce que le _if_ teste. Ça rend la lecture de la méthode principale plus facile :p

Aussi, dans la 3e section de la condition (après le 2e _or_), on compare un index (_self.last_altitude_index_) avec une valeur d'altitude (_values[-1]_). Est-ce que c'est voulu? Si oui, pourquoi on fait ça?

BaboucheOne replied:
Parle tu du second if ? ou des deux if ?

MaitreMenard replied:
Celui avec la très longue expression à la ligne 19.

MaitreMenard commented on BaseStation/src/consumer.py:
```diff
@@ -36,10 +39,17 @@ def update(self):
                     self.data[key].append(value)
                 self.data["altitude_feet"].append(packet.altitude * METERS2FEET)
                 self.manage_coordinates(packet)
+            self.manage_apogee(self.data["altitude_feet"])
```
C'est une bonne idée d'avoir mis le calcul de l'apogée en dehors de la boucle de traitement des packets afin de faire le calcul qu'une seule fois ! :+1:

MaitreMenard commented on BaseStation/src/ui/altitude_graph.py:
```diff
@@ -19,30 +19,34 @@ def __init__(self, parent: QtWidgets.QWidget):
+
+        super(AltitudeGraph, self).paintEvent(ev)
+
     def draw(self, values: list):
```
À renommer _draw_altitude_curve_ pour bien distinguer ce que fait cette méthode par rapport à _draw_apogee_.

MaitreMenard commented on BaseStation/src/consumer.py:
```diff
@@ -36,10 +39,17 @@ def update(self):
 
+    def manage_apogee(self, values: list):
+        self.apogee_calculator.update(values)
+        if self.apogee_calculator.has_apogee():
```
Au lieu d'avoir une méthode _has_apogee_ et d'utiliser les attributs internes de la classe _ApogeeCalculator_, je crois que ça serait préférable d'avoir une méthode _get_apogee_ qui retourne l'apogée sous la forme d'un tuple (x, y) -> (index, altitude) et de simplement assigner ce tuple au champ du consommateur (qui ne sera plus une liste). S'il n'y a pas d'apogée, la méthode retourne _None_, mais le consommateur ne s'en préoccupe pas. Dans _AltitudeGraph_, il n'y a qu'à remplacer la vérification de la longueur de la liste pour un None check.

MaitreMenard commented on BaseStation/tests/test_apogee.py:
```diff
@@ -0,0 +1,53 @@
+        self.apogee_calculator.update(points)
+        self.assertEqual(self.apogee_calculator.apogee, 804)
+
+        points = [0, 100, 194, 256, 500, 804, 300, 1000, 600, 9]
```
Normalement, on veut qu'un test unitaire ne teste qu'une seule chose. Comme ça, s'il échoue, on sait tout de suite pourquoi. Une façon simple de voir cela est qu'en général, on veut un seul assert par test. Dans ce cas-ci, il serait préférable de séparer _test_update_ et _test_has_apogee_ en plusieurs tests plus petits avec des noms indiquant clairement le but du test.

MaitreMenard commented on BaseStation/src/ui/altitude_graph.py:
```diff
@@ -19,30 +19,34 @@ def __init__(self, parent: QtWidgets.QWidget):
+        self.apogee_point = self.plotItem.scatterPlot([], [], pxMode=True, size=9, brush=mkBrush(color='b'))
+        self.addItem(self.apogee_text)
+
+    def paintEvent(self, ev):
```
Ça règle le bug qu'on avait avec le TextItem et c'est beaucoup plus clean que ce que j'avais fait! :+1:

MaitreMenard commented:
C'est du bon travail! Les commentaires que j'ai mis plus haut sont surtout des petites améliorations de formes.

Par contre, j'ai trouvé deux bugs en testant la fonctionnalité. Tout d'abord, en mode temps réel, quand on n'a pas de données avant de commencer l'acquistion, les axes s'emballent et changent en permanence d'échelle. Si j'attends assez longtemps, ça fini par faire planter PyQtGraph et notre logiciel arrête de répondre. Si je me dépêche à lancer l'acquisition avant que ça plante, le phénomène arrête dès qu'on reçoit des données et ça marche bien après. Il faudrait vraiment régler ça avant de merger.

Ensuite, en mode replay, si on se met à reculer après avoir dépassé l'apogée, le point bleu de l'apogée reste 1 frame de plus après le point correspondant sur la courbe soit parti. Ce n'est pas si grave et on peut le laisser aller avec le merge, mais si tu trouve comment régler ça, ça serait un plus.
![bug_apogee0](https://user-images.githubusercontent.com/14338925/50362669-90f72680-0536-11e9-8890-b2868655d098.PNG)
![bug_apogee](https://user-images.githubusercontent.com/14338925/50362526-10d0c100-0536-11e9-939c-ca22161f97b3.PNG) 
![bug_apogee2](https://user-images.githubusercontent.com/14338925/50362613-65743c00-0536-11e9-8f63-d269ac4aa09f.PNG)

BaboucheOne committed:
Errors fix

+ Renommer les fonctions
+ Modification de la class 'apogee_calculator'
+ Resolution du bug du mode 'temps reel'

MaitreMenard commented on BaseStation/src/apogee_calculator.py:
```diff
@@ -0,0 +1,50 @@
+        else:
+            self.set_value_none()
+
+        self.last_altitude = values[length-2]
```
Pourquoi est-ce qu'on prend l'avant-dernière valeur au lieu de la dernière? Aussi, à quoi sert l'attribut _last_altitude_? Il ne semble pas être utilisé.

MaitreMenard committed:
Fix changement d'échelles des axes du graphique d'altitude

BaboucheOne committed:
Cleanup du code

BaboucheOne committed:
Details Fix

+ De nouveaux tests unitaires pour 'Apogee_calculator'

+ Changement de l'UI en mode 'real time'

BaboucheOne committed:
Update altitude_graph.py

BaboucheOne committed:
Modifications des test unitaires

BaboucheOne committed:
Amelioration des tests unitaires
